### PR TITLE
[TSI] Implement Get APIs for Types

### PR DESCRIPTION
**Pull request by bikamani**
Changes include the following

- Implementation for Get and Get Batch APIs
- Few design changes and renaming

This PR does not include e2e tests, will be added in follow up PRs after rest of the implementation.

### CODE REVIEW

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/design/TimeSeriesTypes_API_Design.md [hunk 1]
```diff
@@ -28,25 +28,25 @@ Executes a batch get, create, update, delete operation on multiple time series t
...
-/// Gets time series insight types by Time Series Id asynchronously.
+/// Gets time series insight types by Time Series Type Ids asynchronously.
 /// </summary>
-/// <param name="timeSeriesIds">List of Ids of the Time Series instances.</param>
+/// <param name="timeSeriesTypeIds">List of Time Series Type Ids of the Time Series Types to return.</param>
 /// <param name="cancellationToken">The cancellation token.</param>
 /// <returns>List of types or error objects corresponding by position to the array in the request. Type object is set when operation is successful and error object is set when operation is unsuccessful.
-public virtual async Task<Response<TimeSeriesTypeOperationResult[]>> GetTimeSeriesTypesAsync(
-    IEnumerable<TimeSeriesId> timeSeriesIds, 
+public virtual async Task<Response<TimeSeriesTypeOperationResult[]>> GetTimeSeriesTypesbyId(
```
Comment (bikamani) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/design/TimeSeriesTypes_API_Design.md:36]: Renaming to add "byId" or "byNames" since both are taking IEnumerable<string>

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/design/TimeSeriesTypes_API_Design.md [hunk 2]
```diff
@@ -28,25 +28,25 @@ Executes a batch get, create, update, delete operation on multiple time series t
...
+    IEnumerable<string> timeSeriesTypeIds,
```
Comment (bikamani) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/design/TimeSeriesTypes_API_Design.md:37]: Kept it IEnumerable for now. Pending discussion with Krzysztof.

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 3]
```diff
@@ -1054,5 +1056,329 @@ public virtual Response<SearchSuggestion[]> GetSearchSuggestions(string searchSt
...
+        /// <remarks>
+        /// For more samples, see <see href="https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/samples">our repo samples</see>.
+        /// </remarks>
+        /// <exception cref="ArgumentNullException">
+        /// The exception is thrown when <paramref name="timeSeriesTypeNames"/> is <c>null</c>.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        /// The exception is thrown when <paramref name="timeSeriesTypeNames"/> is empty.
+        /// </exception>
+        public virtual async Task<Response<TimeSeriesTypeOperationResult[]>> GetTimeSeriesTypesAsyncbyNames(
```
Comment (bikamani) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:1196]: Here and below are batch Get APIs.

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/design/TimeSeriesTypes_API_Design.md [hunk 4]
```diff
@@ -28,25 +28,25 @@ Executes a batch get, create, update, delete operation on multiple time series t
...
-/// Gets time series insight types by Time Series Id asynchronously.
+/// Gets time series insight types by Time Series Type Ids asynchronously.
 /// </summary>
-/// <param name="timeSeriesIds">List of Ids of the Time Series instances.</param>
+/// <param name="timeSeriesTypeIds">List of Time Series Type Ids of the Time Series Types to return.</param>
 /// <param name="cancellationToken">The cancellation token.</param>
 /// <returns>List of types or error objects corresponding by position to the array in the request. Type object is set when operation is successful and error object is set when operation is unsuccessful.
-public virtual async Task<Response<TimeSeriesTypeOperationResult[]>> GetTimeSeriesTypesAsync(
-    IEnumerable<TimeSeriesId> timeSeriesIds, 
+public virtual async Task<Response<TimeSeriesTypeOperationResult[]>> GetTimeSeriesTypesbyIdAsync(
```
Comment (bikamani) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/design/TimeSeriesTypes_API_Design.md:36]: Renaming to add "byId" or "byNames" since both are taking IEnumerable

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 5]
```diff
@@ -1054,5 +1056,329 @@ public virtual Response<SearchSuggestion[]> GetSearchSuggestions(string searchSt
...
+            }
+            catch (Exception ex)
+            {
+                scope.Failed(ex);
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Gets Time Series types by Time Series Types Names asynchronously.
```
Comment (bikamani) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:1179]: Here and below are batch Get APIs.

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 6]
```diff
@@ -1054,5 +1056,329 @@ public virtual Response<SearchSuggestion[]> GetSearchSuggestions(string searchSt
...
+        /// <returns>The pageable list <see cref="AsyncPageable{TimeSeriesType}"/> of Time Series instances types with the http response.</returns>
+        public virtual AsyncPageable<TimeSeriesType> GetTimeSeriesTypesAsync(
+            CancellationToken cancellationToken = default)
+        {
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope($"{nameof(TimeSeriesInsightsClient)}.{nameof(GetTimeSeriesTypes)}");
+            scope.Start();
+
+            try
+            {
+                async Task<Page<TimeSeriesType>> FirstPageFunc(int? pageSizeHint)
```
Comment (timtay-microsoft) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:1073]: I don't see the pageSizeHint being used anywhere. Does the service not support it?

**Review by timtay-microsoft — APPROVED** [review]

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs (hunk 6 above)
Comment (timtay-microsoft) [reply sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:1073]: Nor do I see it used in the nextPageFunc

Comment (bikamani) [reply sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:1073]: That's a very good observation, I'll follow up with Tsi team and update this code accordingly.

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 7]
```diff
@@ -1054,5 +1056,329 @@ public virtual Response<SearchSuggestion[]> GetSearchSuggestions(string searchSt
                 throw;
             }
         }
+
+        /// <summary>
+        /// Gets time series insight types in pages asynchronously.
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:1061]: Might want to confirm with Eric, but I believe he wanted 'time series insights' to be always cased as 'Time Series Insights'

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs (hunk 5 above)
Comment (barustum) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:1179]: Maybe reword to 'Gets Time Series Insights types...', similar to previous methods.

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 8]
```diff
@@ -1054,5 +1056,329 @@ public virtual Response<SearchSuggestion[]> GetSearchSuggestions(string searchSt
...
+            {
+                scope.Failed(ex);
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Gets Time Series types by Time Series Types Names asynchronously.
+        /// </summary>
+        /// <param name="timeSeriesTypeNames">List of names of the Time Series Types to return.</param>
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:1181]: not sure if 'Types' need to be upper case.

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 9]
```diff
@@ -1054,5 +1056,329 @@ public virtual Response<SearchSuggestion[]> GetSearchSuggestions(string searchSt
...
+            }
+            catch (Exception ex)
+            {
+                scope.Failed(ex);
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Gets Time Series types by Time Series Types Names synchronously.
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:1231]: similar comment as above

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 10]
```diff
@@ -1054,5 +1056,329 @@ public virtual Response<SearchSuggestion[]> GetSearchSuggestions(string searchSt
...
+        /// </summary>
+        /// <param name="timeSeriesTypeNames">List of names of the Time Series Types to return.</param>
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:1233]: We might want to check if Types need to be upper case. Probably not. Maybe you can check out the TSI MS documentation pages to confirm how they case it.

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 11]
```diff
@@ -1054,5 +1056,329 @@ public virtual Response<SearchSuggestion[]> GetSearchSuggestions(string searchSt
...
+        /// The exception is thrown when <paramref name="timeSeriesTypeNames"/> is <c>null</c>.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        /// The exception is thrown when <paramref name="timeSeriesTypeNames"/> is empty.
+        /// </exception>
+        public virtual async Task<Response<TimeSeriesTypeOperationResult[]>> GetTimeSeriesTypesbyNamesAsync(
+            IEnumerable<string> timeSeriesTypeNames,
+            CancellationToken cancellationToken = default)
+        {
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope($"{nameof(TimeSeriesInsightsClient)}.{nameof(GetTimeSeriesTypes)}");
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:1200]: nameof(GetTimeSeriesTypesbyNames)

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 12]
```diff
@@ -1054,5 +1056,329 @@ public virtual Response<SearchSuggestion[]> GetSearchSuggestions(string searchSt
...
+        /// <remarks>
+        /// For more samples, see <see href="https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/samples">our repo samples</see>.
+        /// </remarks>
+        /// <exception cref="ArgumentNullException">
+        /// The exception is thrown when <paramref name="timeSeriesTypeNames"/> is <c>null</c>.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        /// The exception is thrown when <paramref name="timeSeriesTypeNames"/> is empty.
+        /// </exception>
+        public virtual async Task<Response<TimeSeriesTypeOperationResult[]>> GetTimeSeriesTypesbyNamesAsync(
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:1196]: I believe the 'by' should be 'By'

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 13]
```diff
@@ -1054,5 +1056,329 @@ public virtual Response<SearchSuggestion[]> GetSearchSuggestions(string searchSt
...
+        /// The exception is thrown when <paramref name="timeSeriesTypeNames"/> is <c>null</c>.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        /// The exception is thrown when <paramref name="timeSeriesTypeNames"/> is empty.
+        /// </exception>
+        public virtual Response<TimeSeriesTypeOperationResult[]> GetTimeSeriesTypesbyNames(
+            IEnumerable<string> timeSeriesTypeNames,
+            CancellationToken cancellationToken = default)
+        {
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope($"{nameof(TimeSeriesInsightsClient)}.{nameof(GetTimeSeriesTypes)}");
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:1252]: nameof(GetTimeSeriesTypesByNames)

File: sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs [hunk 14]
```diff
@@ -1054,5 +1056,329 @@ public virtual Response<SearchSuggestion[]> GetSearchSuggestions(string searchSt
...
+        /// The exception is thrown when <paramref name="timeSeriesTypeIds"/> is <c>null</c>.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        /// The exception is thrown when <paramref name="timeSeriesTypeIds"/> is empty.
+        /// </exception>
+        public virtual async Task<Response<TimeSeriesTypeOperationResult[]>> GetTimeSeriesTypesbyIdAsync(
+            IEnumerable<string> timeSeriesTypeIds,
+            CancellationToken cancellationToken = default)
+        {
+            using DiagnosticScope scope = _clientDiagnostics.CreateScope($"{nameof(TimeSeriesInsightsClient)}.{nameof(GetTimeSeriesTypes)}");
```
Comment (barustum) [inline sdk/timeseriesinsights/Azure.Iot.TimeSeriesInsights/src/TimeSeriesInsightsClient.cs:1303]: nameof(GetTimeSeriesTypesById)
